src/java/org/apache/cassandra/tools/NodeTool.java

```diff
@@ -46,7 +47,20 @@
 {
     private static final String HISTORYFILE = "nodetool.history";
 
+    private static NodeProbeFactory nodeProbeFactory = new NodeProbeFactory();
```

Comment:
Rather than rely on mutable state, can we just have the CLI and core implementation decoupled?  CLI can be simply a ```new NodeProbeFactory.execute(args)``` and the in-jvm dtests use a different implementation?

Reply:
or make execute a instance method and provide this in the `NodeTool` constructor, so main is `new NodeTool().execute(args)` and dtest is `new NodeTool(override).execute(args)`

Reply:
The `NodeToolCmd` depends on `NodeProbe`. In order to inject the mock dependency, we either passing the mock instance to the cmd instances or like what the patch current have, depends on a static field. Why the `nodeProbeFactory` cannot be an instance field? Because the `NodeToolCmd` is a static inner class, it cannot see unless the field is a class field. 

If passing the mock instance reference, we will end up having the following code in the `NodeTool#execute` method, which I think is ugly. The reason of injecting dependency by doing the type check and casting is that cli parses command into runnable. 

```
            Runnable parse = parser.parse(args);
            if (parse instanceof NodeToolCmd) // hmmm....
            {
                ((NodeToolCmd) parse).setNodeProbeFactory(nodeProbeFactory); // or pass in the probe reference
            }
            printHistory(args);
            parse.run();
```

Reply:
with a small change, you can get this without statics.

by replacing our usage of `Runnable` with `Consumer<NodeProbeFactory>` (need to extend `Help` to do this), we can replace `command.run()` with `command.accept(nodeProbeFactory)`.  Since the usage is just the `connect` method, this won't change the different commands, and will be local to `NodeToolCmd`